Druid/DRCogitCodeGenerator.class.st

```diff
@@ -266,15 +266,24 @@ DRCogitCodeGenerator >> generateCompare: aComparison operands: operands [
 { #category : #'generate-code' }
 DRCogitCodeGenerator >> generateCompileTimeIfCondition: aDRBranchIfCondition [
 
-	| jitCondition sequenceNode |
-	jitCondition := aDRBranchIfCondition conditionAsJitCompileTimeExpression.
+	aDRBranchIfCondition isConditionalLoop ifTrue: [ ^ self generateLoop: aDRBranchIfCondition asConditionalLoop ].
+
+	self generateConditional: aDRBranchIfCondition.
 
-	sequenceNode := self innerCompilePathFrom: aDRBranchIfCondition trueBranch.
+
+]
+
+{ #category : #'code-generation' }
+DRCogitCodeGenerator >> generateConditional: aDRBranchIfCondition [
+
+	| jitCondition innerGenerator |
+	jitCondition := aDRBranchIfCondition conditionAsJitCompileTimeExpression.
+	innerGenerator := self innerCompilePathFrom: aDRBranchIfCondition trueBranch.
```

Comment:
Crazy idea: what if we do:

```smalltalk
stagedTrueBranch := self innerCompileBlocks: aDRBranchIfCondition trueBranch.
stagedFalseBranch := self innerCompileBlocks: aDRBranchIfCondition falseBranch.
```

And then we generate an ifTrue:ifFalse: ?
Worst case, one of the branches will be an empty block, right?

Reply:
The problem is not the empty block but the phis after the if.

Here you also need to calculate the path from each branch until the joint point.

I think that the problem includes the result of the phi as a jit expression. In that case, we need to store variable(s) similar to the inductive variables for the loop in this PR.

I'm not sure what is better...